### ISSUE TITLE
SimGCL在论文中处于underfitting的状态

### ISSUE THREAD

**Issue by Coder-Yu** [HKUDS/LightGCL#6]
作者你好，我是SimGCL的作者，经常有关注贵组的工作。最近也有打算复现LightGCL。

但我发现在LightGCL的论文中，SimGCL似乎完全处于欠拟合或者错误选择超参数的状态。我使用了本repo里提供的yelp和gowalla数据集重新对SimGCL进行了测试。在保持general setting与文中一致的情况下，我凭经验随机选择了 lambda_cl = 0.2, epsilon=0.1, tau=0.2的组合 （后两个值对大多数数据集来说是SimGCL的相对最优超参），在迭代所有训练样本第二次之后的结果即超过了LightGCL的结果，也远超SimGCL在文中的结果。我的结果如下：

Yelp: SimGCL 第二次迭代            Recall@20:   0.0962            NDCG@20:   0.0833
Yelp: SimGCL 收敛（第九次）     Recall@20:   **0.1048**            NDCG@20:   **0.0903**
yelp: SimGCL 论文汇报结果         Recall@20:   0.0718            NDCG@20:   0.0615
yelp: 文中LightGCL结果         Recall@20:   0.0793            NDCG@20:   0.0668

Gowalla: SimGCL 第二次迭代       Recall@20:   0.1739            NDCG@20:   0.1060
Gowalla: SimGCL 收敛 （第十次）Recall@20:   **0.1893**            NDCG@20:   **0.1145**
Gowalla: SimGCL 论文汇报结果    Recall@20:   0.1357            NDCG@20:   0.0818
Gowalla: 文中LightGCL结果         Recall@20:   01578            NDCG@20:   0.0935

文中提到 “To ensure a fair comparison, we tune the hyperparameters of all the baselines within the ranges suggested in the original papers.” 但我发现贵组实验中实际上可能使用了 lambda_cl = 0.01, tau=0.1。 lambda_cl = 0.01在SimGCL的文章参数敏感性实验中已被表明为三个数据集上较差的选择。另外SimGCL文章中也提到“ In SimGCL and SGL, we empirically let the temperature 𝜏 = 0.2, and this value is also reported as the best in the original paper of SGL.” tau这个参数实际上是较为敏感的，我的经验是0.2变到0.1会出现结果的较大波动。似乎LightGCL实验时并未参考SimGCL的文章。

以上关于SimGCL的结果均为通过[SELFRec](https://github.com/Coder-Yu/SELFRec)得到。感兴趣的话可以对比是否我们关于SimGCL的实现有不同之处，导致了论文里面的问题。

------------------------- UPDATE---------------------------------------------------------------------

我用lambda_cl = 0.2, epsilon=0.1, tau=0.2的组合在贵组的SSLRec上尝试了一下，yelp dataset上经过两次迭代后为

Recall@20: 0.0929 NDCG@20: 0.0791

我没有跑完掐掉了，但两次迭代也好于文中汇报的SimGCL与LightGCL的结果。 

然后我也使用了SSLRec里面默认的SimGCL参数  lambda_cl = 0.01, epsilon = 0.2, tau=0.1, 三次迭代记录如下：

{'optimizer': {'name': 'adam', 'lr': 0.001, 'weight_decay': 0}, 'train': {'epoch': 100, 'batch_size': 256, 'save_model': False, 'loss': 'pairwise', 'test_step': 1}, 'test': {'metrics': ['recall', 'ndcg'], 'k': [10, 20], 'batch_size': 256}, 'data': {'type': 'general_cf', 'name': 'yelp', 'user_num': 29601, 'item_num': 24734}, 'model': {'name': 'simgcl', 'keep_rate': 1.0, 'layer_num': 2, 'reg_weight': 1e-06, 'cl_weight': 0.01, 'temperature': 0.1, 'embedding_size': 32, 'eps': 0.2}, 'tune': {'enable': False}, 'device': 'cuda'}
Training Recommender: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4177/4177 [03:48<00:00, 18.26it/s]
[Epoch   0 / 100] bpr_loss: 0.2147 reg_loss: 0.0304 cl_loss: 0.1139 
[recall@10: 0.0439 recall@20: **0.0728** ] [ndcg@10: 0.0533 ndcg@20: **0.0621** ] 
Training Recommender: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4177/4177 [03:48<00:00, 18.29it/s]
[Epoch   1 / 100] bpr_loss: 0.1037 reg_loss: 0.0536 cl_loss: 0.1022 
[recall@10: 0.0478 recall@20: **0.0810** ] [ndcg@10: 0.0582 ndcg@20: **0.0686** ] 
Training Recommender: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4177/4177 [03:48<00:00, 18.28it/s]
[Epoch   2 / 100] bpr_loss: 0.0929 reg_loss: 0.0623 cl_loss: 0.0932 
[recall@10: 0.0499 recall@20: **0.0840** ] [ndcg@10: 0.0605 ndcg@20: **0.0711** ] 

第二次迭代完之后也已经超过了LightGCL的结果。也可以看出这组参数确实差于SimGCL论文里推荐选择的参数。

**Comment by akaxlh** [comment]
Junliang你好，我是LightGCL的作者之一，贵组的SimGCL也是最近半年来我们组最为关注、讨论最多的工作之一，很高兴有机会和您探讨SSL for Rec相关话题。

在LightGCL工作中，我们采用了一种先采样user、再采样user interaction的方式进行训练样本生成，相比于现在我们所采用的interaction采样，这种方式会普遍地导致所有模型训练不充分，也造成了您的实验中SimGCL相对我们论文中的结果显著提高。

LightGCL这篇工作完成于去年年中，我们从去年下半年开始了<a href='https://github.com/HKUDS/SSLRec'>SSLRec</a>框架的建设，并在之后的工作中采用了这一代码框架，或每个同学基于这一代码框架的个性化版本。

我们注意到您和一些其他关注者因为代码框架差异对实验结果产生的困惑，所以在SSLRec基础上，参照SimGCL参数对LightGCL进行了重新训练，感谢您提供的实验结果，我们得以即时展示以下结果：

| Dataset | Metric | SimGCL | LightGCL-dim64 | LightGCL-dim32  | LightGCL-dim32-gnn1|
|:-------:|:--------:|:--------:|:--------:|:--------:|:--------:|
|Yelp   |Recall@20 | 0.1048 | 0.0995 $\pm$ 0.0010 | 0.0961 $\pm$ 0.0010| 0.0990 |
|Yelp   |NDCG@20   | 0.0903 | 0.0857 $\pm$ 0.0009 | 0.0826 $\pm$ 0.0007| 0.0852 |
|Gowalla|Recall@20 | 0.1893 | 0.2122 $\pm$ 0.0016 | 0.1920 $\pm$ 0.0018 | 0.2001 |
|Gowalla|NDCG@20   | 0.1145 | 0.1238 $\pm$ 0.0013 | 0.1110 $\pm$ 0.0012| 0.1176 |

可以观察到，LightGCL与SimGCL在不同指标、数据集上互有优劣，这验证了LightGCL与代码实现无关的有效性。新的代码和相关文档已更新至本repo中，欢迎您对代码进行测试，如有问题请一定联系我们。

**Comment by Coder-Yu** [comment]
Hi, 感谢回复。

Gowalla上我重新选择了lambda_cl=0.5，其他参数保持不变的情况下两层SimGCL-dim64的效果如下：

7次迭代之后 Recall@20: 0.22158   |   NDCG@20: 0.13546

个人觉得使用baseline原文没用过的数据集的时候，出现baseline没有调好的情况还是挺常见的。要求每个baseline都调的跟原作上场一样好对于任何研究人员都是求全责备，能接近optimal就可以了。建议一作可以重新做一下LightGCL文章里面的实验，对baseline稍细致的调参一下然后再比较，最后更新一个arxiv的版本。相信大多数读者以及作者都会觉得是一个可接受的方案。

**Comment by akaxlh** [comment]
感谢回复。我们会在后续更新相关内容。
